etcd-recovery/etcdrecovery.go

```diff
@@ -0,0 +1,538 @@
+package etcdrecovery
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"os"
+	"os/signal"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/spf13/cobra"
+	"go.uber.org/zap/zapcore"
+
+	hyperapi "github.com/openshift/hypershift/support/api"
+	hyperutil "github.com/openshift/hypershift/support/util"
+	"go.etcd.io/etcd/api/v3/etcdserverpb"
+	"go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	etcdclient "go.etcd.io/etcd/client/v3"
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	crclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/log/zap"
+)
+
+const (
+	defaultEtcdClientTimeout = 5 * time.Second
+	defaultDialTimeout       = 3 * time.Second
+	healthyClusterWait       = 5 * time.Minute
+)
+
+type options struct {
+	etcdClientCertFile string
+	etcdClientKeyFile  string
+	etcdCAFile         string
+	namespace          string
+}
+
+func NewRecoveryCommand() *cobra.Command {
+	opts := options{
+		etcdClientCertFile: "/etc/etcd/tls/client/etcd-client.crt",
+		etcdClientKeyFile:  "/etc/etcd/tls/client/etcd-client.key",
+		etcdCAFile:         "/etc/etcd/tls/etcd-ca/ca.crt",
+	}
+
+	cmd := &cobra.Command{
+		Use:          "recover-etcd",
+		Short:        "Commands to report on etcd status and recover member that is not functional",
+		SilenceUsage: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			cmd.Help()
+			os.Exit(1)
+		},
+	}
+
+	cmd.PersistentFlags().StringVar(&opts.etcdClientCertFile, "etcd-client-cert", opts.etcdClientCertFile, "etcd client cert file.")
+	cmd.PersistentFlags().StringVar(&opts.etcdClientKeyFile, "etcd-client-key", opts.etcdClientKeyFile, "etcd client cert key file.")
+	cmd.PersistentFlags().StringVar(&opts.etcdCAFile, "etcd-ca-cert", opts.etcdCAFile, "etcd trusted CA cert file.")
+	cmd.PersistentFlags().StringVar(&opts.namespace, "namespace", "", "namespace of etcd cluster")
+	_ = cmd.MarkFlagRequired("etcd-endpoints")
+	_ = cmd.MarkFlagRequired("namespace")
+
+	cmd.AddCommand(NewStatusCommand(&opts))
+	cmd.AddCommand(NewRunCommand(&opts))
+
+	return cmd
+}
+
+func NewRunCommand(opts *options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:          "run",
+		Short:        "Recover etcd cluster that has quorum",
+		SilenceUsage: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx, cancel := signal.NotifyContext(cmd.Context(), os.Interrupt, syscall.SIGINT)
+			defer cancel()
+			ctx = setupCmdContext(ctx)
+			err := runRecovery(ctx, *opts)
```

Comment:
run recovery might fail for transient issues or legitimately, in any case we fail here which kill the pod and hide logs?

Reply:
The job pod terminates and a new one is launched. All job pods remain around until we exhaust retries (10) or one succeeds.